동기화 메모장 [STEP 2] Marco

Marcorable opened the pull request:
@SungPyo 

@keeplo

안녕하세요 웨더!🤠

기능의 깊이를 모두 공부하고 구현 했다고 할 수 없어서 조금 아쉽지만 
프로젝트에서 요구되는 기능을 구현하는 수준까지 공부를 해보고 진행해 보았습니다.

그리고 Step1에서 구현 했었던 부분을 가지고도 진행하다가 고민하기도 하고 변경을 했다가 다시 바꾸기도 하고
많은 변화가 있었습니다.. 완전 다른 코드 같더라도 이해해주세요 ㅎㅎ;; 

## Step1 에서 바뀐부분..
앞서 말씀드린 것처럼 Step1과 구현 방향을 조금 바꾸었습니다.
<img width="385" alt="image" src="https://user-images.githubusercontent.com/24707229/133258208-4ae97681-be77-4489-b959-175f0793b912.png">
먼저 UITableViewDataSource나 TextViewDelegate를 객체로 생성하는 방향은 동기화 메모앱에서 
한 테이블로 다른 두 화면을 보여준다거나 여러 편집 객체가 있어서 각 UIElements에 동작을 구현해놓기 보다
ViewController에서 UI Elements의 동작을 전반적으로 모두 구현해도 나쁘지 않겠다고 생각해서 바꾸었습니다. 

메모 선택을 위한 테이블 뷰를 그려주고 앱 전반에서 사용될 메모 인스턴스의 배열의 위치를 고민해보았습니다.
<img width="535" alt="image" src="https://user-images.githubusercontent.com/24707229/133258373-9992f221-75ac-4dd6-bd83-9849f67d0591.png">
<img width="757" alt="image" src="https://user-images.githubusercontent.com/24707229/133258452-23e9246c-61d9-4651-9f68-214bf05f714b.png">

다양하게 접한 책 예제나 몇몇 블로그에서 위 코드 처럼 AppDelegate.swift 에 배열을 선언하고 해당 배열을 사용하는 
ViewController에서 AppDelegate의 delegate 형태로 접근 하는 형태를 많이 보았는 데. 
이런 방향도 괜찮은 지도 궁금합니다..
저는 위 방향에서 강제 언래핑을 하지 않고 사용하기 너무 어렵고 동기화 메모장이라면 괜찮지만 기왕이면 Split View 묶음 (primary, secondary) 안에서만 사용되는 데이터라고 생각해서 SplitView에 두고 초기화단계에서 delegate를 전달해 보았습니다.

## 고민되었던 점 해결하지 못한 부분 🤯
* **NavigationItem의 동작 불능**
저는 많은 동작이나 기능에 대해 애플 기본 메모앱을 많이 참고 했는데 기능 요구서에 없었지만 구현한 부분은 
TextView의 상태가 Editing으로 전환되고 키보드가 올라오는 상태에서 rightBarButton으로 **Done**이 생성되고
Done 버튼을 누르면 키보드가 내려가능 동작을 하는 구현을 해보았습니다. 

고민하다 아래 코드처럼 구현한 이유는 Editing 상태일때와 아닐때 **Done** 버튼의 위치 변화를 애플 메모와 동일하게 해보고 싶었습니다.. 
~~~swift
// SecondaryViewController

class SecondaryViewController {
    private let hidableDoneButton = UIBarButtonItem(barButtonSystemItem: .done,
                                                    target: nil,
                                                    action: #selector(resignFromTextView))
    private let seeMoreStaticButton = UIBarButtonItem(image: UIImage(systemName: "ellipsis.circle"),
                                                      style: .plain,
                                                      target: nil,
                                                      action: #selector(tappingSeeMoreButton))
    // MARK: - Keyboard Notification
    @objc func keyboardWasShown(_ notification: Notification) {
        setBarButtons(isHide: false)
    }
    @objc func keyboardWillBeHidden(_ notification: Notification) {
        setBarButtons(isHide: true)
    }
    func setBarButtons(isHide: Bool) {
        self.navigationItem.rightBarButtonItems = isHide ? [seeMoreStaticButton] : [hidableDoneButton, seeMoreStaticButton]
    }
}

}
~~~
![rightBarButton](https://user-images.githubusercontent.com/24707229/133257355-4b7b0d33-06a7-40ee-b8cf-aa25401e693b.gif)
화면에서 두 버튼이 상황에 맞게 잘 등장해서 잘 동작할 줄 알았는 데. 이상하게 TextView가 firstResponder가 되지 않는 (키보드가 숨겨진 상황)에서 더보기 버튼만 있을때는 버튼이 동작하지 않고 반대 상황에서 두 버튼은 모두 원하는 데로 동작이 됐습니다..
어떤 부분에서 문제가 있는지 해결하지 못했습니다.


## 궁금한 점 🤔
* **테이블 뷰 셀 동적인 셀 크기 지정**
<img width="1694" alt="테이블뷰 자동 높이1" src="https://user-images.githubusercontent.com/24707229/133260036-4df01b4d-a745-46d2-a987-941913cfedd3.png">
<img width="1702" alt="테이블뷰 자동 높이2" src="https://user-images.githubusercontent.com/24707229/133260111-1652ba31-5cb2-4da8-86ad-110f6bacc8ea.png">

야곰 블로그의 [AutoLayout 정복하기](https://yagom.net/courses/autolayout/lessons/working-with-self-sizing-table-view-cells/topic/the-basis-2/)를 참고하고 다양한 블로그를 보니 셀 content 내부의 객체들의 높이가 정확하게 정해지면 셀의 높이가 자동으로 맞춰진다고 이해를 했습니다.  
위 사진처럼 warning을 없애는 방법으로 초기 높이를 지정해줘서 없앤다고 했는데 
이런 현상을 옳게 이해한건지 혹시 위 코드에서 문제는 없는 지 궁금합니다..


## 피드백 받고 싶은 부분 😹
* **TableView 화면 ViewController가 ViewWillAppear 될대마다 CoreData fetch 동작**
```swift
override func viewWillAppear(_ animated: Bool) {
    super.viewWillAppear(animated)
    print("PrimaryViewController - viewWillAppear")
    rootDelegate.listResource = coreManager.fetchData()
    if rootDelegate.listResource.isEmpty {
        NSLog("CoreData 데이터 없어서 asset 데이터(Sample) 추가")
        rootDelegate.listResource = rootDelegate.readDataAsset()
        for data in rootDelegate.listResource {
            coreManager.insertData(data)
        }
    }
    tableView.reloadData()
}
```
위 코드는 메모앱 리스트를 보여주는 ViewController의 모습입니다.
리스트가 아닌 디테일 뷰에서 메모가 삭제 또는 수정 되는 경우 SplitView를 통해 CoreData와 앱에서 사용되는 메모 배열에 데이터를
삭제를 하도록 구현했습니다.

이렇게 구현한 이유는 primary 뷰와 secondary 뷰 간에 메시지 전달을 SplitView를 통해서 하도록 하려고 했고
일일이 각 뷰까지 도달하기 보다 각 뷰가 CoreData 접근할때 업데이트된 데이터를 fetch해서 실시간으로 변화를 알 수 있도록 하려고 
구현해보았습니다.

페이지를 이동할때마다 CoreData의 fetch 동작을 진행하는데 혹시 이런 형태의 구현에 리스크는 없을지 피드백 받고 싶습니다.

 * **두 화면이 모두 보일때 동작**
![Simulator Screen Recording - iPhone 12 Pro Max - 2021-09-14 at 22 01 17](https://user-images.githubusercontent.com/24707229/133262626-af2bbc64-971a-456c-aaba-2cb06db5cd43.gif)

아직 해결되지 않은 문제는 + 버튼을 눌러서 새로운 메모 셀을 만들어 두고 DetailView에서 아무런 입력없이 
다른 셀을 선택하면 새로운 메모 셀이 사라지는 동작이 구현하지 못했습니다.
해당 문제가 디테일뷰의 `func textViewDidEndEditing(_ textView: UITextView)` delegate 를 기점으로 테이블 데이터를 
업데이트 하는데 FullScreen으로 한 화면씩 보이는 경우에서는 뒤로가기 탭으로 테이블뷰로 나오면
원하는 데로 되는데  위 사진처럼 할때만 생기는 문제였습니다..

이렇다보니 제가 구현한 형태로 일일이 업데이트되는 시점을 찾고 테이블뷰가 refresh 되게 만드는 방향이 맞는건지..
스플릿 뷰에서 제가 놓친 기능이 있거나 다른 형태로 나누어진 View간에 통신 방법은 없는지 궁금합니다..

혹시 NotificationCenter를 각 뷰간에 통신의 방법으로 사용하는 점도 어떤지 궁금합니다..
저희가 진행했던 쥬스메이커 만들기 공부할때 NotificationCenter 사용은 해당 프로젝트 상황(모델의 데이터 수정하는 요청)에 어울리지 않다고 피드백을 받았는데,
그러면 주로 어떤 상황에서 Notification Center가 사용되는 지 알고 싶습니다.

* * *
사실 코드의 모든 부분을 질문을 하고 싶을 정도로 혼자 하다보니 다양한 선택과 동작 구현에만 초점이 잡힌 것 같습니다..
사소한 부분이라도 이상한 구현이나 잘못된 부분이 있다면 말씀해주세요 ㅎㅎ;;

잘부탁드립니다!

Marcorable committed:
refactor: Summary 문자열 만들기 변경

Marcorable committed:
feat: TextView의 커서가 생길때 키보드의 동작

TextView가 FirstResponder로 커서가 생길때 키보드를 숨기고
resignResponder 동작하는 경우를 구현

Marcorable committed:
refactor: 타입의 이름을 identifier로 가지는 extension 추가

Marcorable committed:
refactor: TableView Resource List 저장 위치 변경

TableViewDataSource -> AppDelegate
접근성, 데이터 보존, 일관성을 위해 Resource 위치를 변경함

Marcorable committed:
refactor: TempMemo 삭제, 화면 구성 제약 변경

Marcorable committed:
feat: 기존 JSON 디코딩 Memo 타입과 CoreData 데이터를 표현하는 MemoData 타입 추가

Marcorable committed:
refactor: 초기화 설정 변경

Marcorable committed:
refactor: 디렉토리 변경

Marcorable committed:
feat: Memo DAO 및 fetch 메서드 구현, .xcdatamodeld 파일 구현

Marcorable committed:
refactor: Key 값을 위한  enum 문자열 반환값 추가

Marcorable committed:
feat: Memorable 모델 타입 구조체로 변경

Marcorable committed:
refactor: 에러처리 필요한 부분 NSLog로 변경 및 출력 문자열 변경

Marcorable committed:
feat: Memo DAO의 insertData메서드 구현

Marcorable committed:
feat: Memo DAO deleteData 메서드 구현

Marcorable committed:
refactor: Memo DAO deleteData메서드 수정

CoreData의 object 불러오기 부분 수정

Marcorable committed:
feat: Memo DAO editData메서드 구현

Marcorable committed:
refactor: CoreData 기본 Container 설정 변경

NSPersistentCloudKitContainer -> NSPersistentContainer

Marcorable committed:
feat: CoreDataManager 싱글톤 형태로 구현

Marcorable committed:
feat: CloudNotes.xcdatamodeld 파일 생성 및 해당 .swift 파일 생성

Entity Name : Memo
class Name : MemoManagedObject

Marcorable committed:
refactor: 문자열 key를 알려주는 enum 타입 제거

Marcorable committed:
refactor: 기존 추상화 프로토콜 이름 Memorable -> MemoModel 로 변경

Marcorable committed:
feat: 메모 완료 버튼 숨기기 기능 추가

Marcorable committed:
feat: 더보기 Alert 기본 동작 구현

Marcorable committed:
refactor: CoreData 문자열 데이터 상수 할당

데이터 접근을 위한 key가 될 문자열 상수처리

Marcorable committed:
refactor : Detail View 업데이트시 SplitView behavior 지정

Detail View 업데이트시 .show 동작 여부 결정하는 Bool 매개변수 추가

Marcorable committed:
feat: 해당 View 내부에서 필요한 문자열 데이터 열거형처리

Marcorable committed:
refactor: 열거형 문자열로 본래 String 데이터 대체

Marcorable committed:
feat: Swipe 동작으로 해당 셀 삭제 동작 구현

해당 셀 유형에 따라 CoreData delete 동작 실행 유무 결정

Marcorable committed:
refactor: 수정된 sample 데이터 삽입 및 변경

Marcorable committed:
feat: activityView 구현 (기초)

Marcorable committed:
refactor: 셀 선택 시 알려주는 방법 변경

데이터만 -> 데이터 + indexPath

Marcorable committed:
feat: 테이블 뷰 셀 swipe 삭제동작 구현

Marcorable committed:
feat: Detail View에서 삭제 동작 구현

Marcorable committed:
refactor: Detail View 구현시 선택된 인덱스 데이터 저장 및 데이터 없을때 빈 페이지 표시 구현

Marcorable committed:
feat: add 동작 구현 및 cell configure  동작 세분화

newMemo -> title : "새로운 메모" / detail: "추가 텍스트 없음" / date: "현재 시간"
기존 Memo -> Memo data 보여주기

Marcorable committed:
refactor: 주석 추가 및 문자열 삭제

Marcorable committed:
refactor: 메모 리스트 위치 변경 appDelegate -> Master SplitView

Marcorable committed:
refactor: MasterView와 통신하는 Delegate 기능 변형

Marcorable committed:
refactor: 변수 사용 방법 변경 및 코드 위치 변경

Marcorable committed:
feat: 앱 전반에서 사용되는 문자열 데이터 Strings enum으로 정리

Marcorable committed:
refactor: 문자열 관련 변경 및 위치조정

문자열 열거형 사용 변경
코드 위치 변경

Marcorable committed:
refactor: TextView의 Delegate 위치 변경

Marcorable committed:
refactor: TextViewDelegate 위치 변경 및 문자열 위치변경

Marcorable committed:
feat: 새로운 메모 추가 또는 메모 수정 동작을 하는 메서드 구현

입력중인 메모의 indexPath의 데이터 objectID가
있으면 수정동작으로 처리
없으면 추가동작으로 처리

Marcorable committed:
refactor: 변수명 변경, 공백처리, print->NSLog 변경 등

Marcorable committed:
refactor: ActivityView 마무리 구현

SungPyo commented:
> @SungPyo
> 
> @Keeplo
> 
> 안녕하세요 웨더!🤠
> 
> 기능의 깊이를 모두 공부하고 구현 했다고 할 수 없어서 조금 아쉽지만
> 프로젝트에서 요구되는 기능을 구현하는 수준까지 공부를 해보고 진행해 보았습니다.
> 
> 그리고 Step1에서 구현 했었던 부분을 가지고도 진행하다가 고민하기도 하고 변경을 했다가 다시 바꾸기도 하고
> 많은 변화가 있었습니다.. 완전 다른 코드 같더라도 이해해주세요 ㅎㅎ;;
> 
> ## Step1 에서 바뀐부분..
> 앞서 말씀드린 것처럼 Step1과 구현 방향을 조금 바꾸었습니다.
> <img alt="image" width="385" src="https://user-images.githubusercontent.com/24707229/133258208-4ae97681-be77-4489-b959-175f0793b912.png">
> 먼저 UITableViewDataSource나 TextViewDelegate를 객체로 생성하는 방향은 동기화 메모앱에서
> 한 테이블로 다른 두 화면을 보여준다거나 여러 편집 객체가 있어서 각 UIElements에 동작을 구현해놓기 보다
> ViewController에서 UI Elements의 동작을 전반적으로 모두 구현해도 나쁘지 않겠다고 생각해서 바꾸었습니다.
👉 `이전에도 취향이라고 말씀드렸지만 개인적으로 저는 이전보다는 지금이 더 좋게 느껴지네요 :)`

> 메모 선택을 위한 테이블 뷰를 그려주고 앱 전반에서 사용될 메모 인스턴스의 배열의 위치를 고민해보았습니다.
> <img alt="image" width="535" src="https://user-images.githubusercontent.com/24707229/133258373-9992f221-75ac-4dd6-bd83-9849f67d0591.png">
> <img alt="image" width="757" src="https://user-images.githubusercontent.com/24707229/133258452-23e9246c-61d9-4651-9f68-214bf05f714b.png">
> 
> 다양하게 접한 책 예제나 몇몇 블로그에서 위 코드 처럼 AppDelegate.swift 에 배열을 선언하고 해당 배열을 사용하는
> ViewController에서 AppDelegate의 delegate 형태로 접근 하는 형태를 많이 보았는 데.
> 이런 방향도 괜찮은 지도 궁금합니다..
👉 `이정도 규모 프로젝트에 정답은 없어요. 하지만 규모가 커지고 모델이 많아진다고 생각하면 정말 끔찍해질 것 같다고 생각해서 저는 절대절대 네버! 지만 말꼬 생각은 어떠신가요? 말꼬 생각이 더 좋을 수 있으니까요.`
> 저는 위 방향에서 강제 언래핑을 하지 않고 사용하기 너무 어렵고 동기화 메모장이라면 괜찮지만 기왕이면 Split View 묶음 (primary, secondary) 안에서만 사용되는 데이터라고 생각해서 SplitView에 두고 초기화단계에서 delegate를 전달해 보았습니다.
👉 `요 부분은 사실 오류가 조금 있는 것 같아요. 동작의 오류라기 보다 네이밍의 오류 라고 해야할까요? delegate가 사실 스플릿 뷰가 들고있는 모델의 이름이지 deleagate기능을 하는게 아니기 때문에(제대로 보지 않았다면 프로토콜인줄 감쪽같이 속았을 겁니다) deleagte라는 이름보다는 model과 관련된 이름이 좋을 것 같아요.`

> ## 고민되었던 점 해결하지 못한 부분 🤯
> * **NavigationItem의 동작 불능**
>   저는 많은 동작이나 기능에 대해 애플 기본 메모앱을 많이 참고 했는데 기능 요구서에 없었지만 구현한 부분은
>   TextView의 상태가 Editing으로 전환되고 키보드가 올라오는 상태에서 rightBarButton으로 **Done**이 생성되고
>   Done 버튼을 누르면 키보드가 내려가능 동작을 하는 구현을 해보았습니다.
> 
> 고민하다 아래 코드처럼 구현한 이유는 Editing 상태일때와 아닐때 **Done** 버튼의 위치 변화를 애플 메모와 동일하게 해보고 싶었습니다..
> 
> ```swift
> // SecondaryViewController
> 
> class SecondaryViewController {
>     private let hidableDoneButton = UIBarButtonItem(barButtonSystemItem: .done,
>                                                     target: nil,
>                                                     action: #selector(resignFromTextView))
>     private let seeMoreStaticButton = UIBarButtonItem(image: UIImage(systemName: "ellipsis.circle"),
>                                                       style: .plain,
>                                                       target: nil,
>                                                       action: #selector(tappingSeeMoreButton))
>     // MARK: - Keyboard Notification
>     @objc func keyboardWasShown(_ notification: Notification) {
>         setBarButtons(isHide: false)
>     }
>     @objc func keyboardWillBeHidden(_ notification: Notification) {
>         setBarButtons(isHide: true)
>     }
>     func setBarButtons(isHide: Bool) {
>         self.navigationItem.rightBarButtonItems = isHide ? [seeMoreStaticButton] : [hidableDoneButton, seeMoreStaticButton]
>     }
> }
> 
> }
> ```
> 
> ![rightBarButton](https://user-images.githubusercontent.com/24707229/133257355-4b7b0d33-06a7-40ee-b8cf-aa25401e693b.gif)
> 화면에서 두 버튼이 상황에 맞게 잘 등장해서 잘 동작할 줄 알았는 데. 이상하게 TextView가 firstResponder가 되지 않는 (키보드가 숨겨진 상황)에서 더보기 버튼만 있을때는 버튼이 동작하지 않고 반대 상황에서 두 버튼은 모두 원하는 데로 동작이 됐습니다..
> 어떤 부분에서 문제가 있는지 해결하지 못했습니다.
👉 ` 증상을 봤지만 굉장히 놀라운 증상이네요. 근데 일단 코드에 문제가 있어요! 만약 화면이 나오자마자 done 버튼을 동작하게 하고싶다고 한다면 done 버튼조차 동작하지 않을거에요 정확한 이유는 BarbuttonItem의 초기화 코드를 봐야 좋겠지만 해당 코드는 애플에서 제공해 주지 않는 부분이라 어느정도 추측만 가능한 상황인데요. 정답만 먼저 알려드리면 let이 아닌 lazy로 선언하고 Target을 self로 지정하면 동작합니다. 그 이유에 대해서는 한번 생각해 보시는것도 좋을 것 같아요! 정 모르시겠다면 다시 질문을 달아주세요 :)`

> ## 궁금한 점 🤔
> * **테이블 뷰 셀 동적인 셀 크기 지정**
> 
> <img alt="테이블뷰 자동 높이1" width="1694" src="https://user-images.githubusercontent.com/24707229/133260036-4df01b4d-a745-46d2-a987-941913cfedd3.png">
> 
> <img alt="테이블뷰 자동 높이2" width="1702" src="https://user-images.githubusercontent.com/24707229/133260111-1652ba31-5cb2-4da8-86ad-110f6bacc8ea.png">
> 
> 야곰 블로그의 [AutoLayout 정복하기](https://yagom.net/courses/autolayout/lessons/working-with-self-sizing-table-view-cells/topic/the-basis-2/)를 참고하고 다양한 블로그를 보니 셀 content 내부의 객체들의 높이가 정확하게 정해지면 셀의 높이가 자동으로 맞춰진다고 이해를 했습니다.
> 위 사진처럼 warning을 없애는 방법으로 초기 높이를 지정해줘서 없앤다고 했는데
> 이런 현상을 옳게 이해한건지 혹시 위 코드에서 문제는 없는 지 궁금합니다..
👉 `음.. 가변적인 높이에 대해서 궁금하신 걸까요..? 원하시는거라면 estimated 에 대해서 공부해 보시면 될 것 같고, 셀 크기를 지정해 주지 않는 경우에는 어느정도의 크기로 보여줘야 하는지 알 수 없기때문에 워닝이 나옵니다 :) 그러니까 이해하신대로 높이를 지정해 주면 워닝이 없어지는게 맞죠! 가변 크기의 경우는 말씀드린대로 estimated 에 대해서 찾아보시면 도움이 될 것 같아요:)`

> ## 피드백 받고 싶은 부분 😹
> * **TableView 화면 ViewController가 ViewWillAppear 될대마다 CoreData fetch 동작**
> 
> ```swift
> override func viewWillAppear(_ animated: Bool) {
>     super.viewWillAppear(animated)
>     print("PrimaryViewController - viewWillAppear")
>     rootDelegate.listResource = coreManager.fetchData()
>     if rootDelegate.listResource.isEmpty {
>         NSLog("CoreData 데이터 없어서 asset 데이터(Sample) 추가")
>         rootDelegate.listResource = rootDelegate.readDataAsset()
>         for data in rootDelegate.listResource {
>             coreManager.insertData(data)
>         }
>     }
>     tableView.reloadData()
> }
> ```
> 
> 위 코드는 메모앱 리스트를 보여주는 ViewController의 모습입니다.
> 리스트가 아닌 디테일 뷰에서 메모가 삭제 또는 수정 되는 경우 SplitView를 통해 CoreData와 앱에서 사용되는 메모 배열에 데이터를
> 삭제를 하도록 구현했습니다.
> 
> 이렇게 구현한 이유는 primary 뷰와 secondary 뷰 간에 메시지 전달을 SplitView를 통해서 하도록 하려고 했고
> 일일이 각 뷰까지 도달하기 보다 각 뷰가 CoreData 접근할때 업데이트된 데이터를 fetch해서 실시간으로 변화를 알 수 있도록 하려고
> 구현해보았습니다.
> 
> 페이지를 이동할때마다 CoreData의 fetch 동작을 진행하는데 혹시 이런 형태의 구현에 리스크는 없을지 피드백 받고 싶습니다.
👉 `음.. 리스크가 있다고 확실하게 대답할 수는 없어요! 정말 viewWillappear 에서 실행해야 할 수도 있기때문에 현재 상태의 프로젝트에서는 다른 사이드가 발생하지 않는 한 문제가 없다고 해도 무리가 없겠네요. 특정 동작이 추가된다거나 하는 경우에는 당연히.. 위치를 옮겨야겠죠?`

> * **두 화면이 모두 보일때 동작**
>   ![Simulator Screen Recording - iPhone 12 Pro Max - 2021-09-14 at 22 01 17](https://user-images.githubusercontent.com/24707229/133262626-af2bbc64-971a-456c-aaba-2cb06db5cd43.gif)
> 
> 아직 해결되지 않은 문제는 + 버튼을 눌러서 새로운 메모 셀을 만들어 두고 DetailView에서 아무런 입력없이
> 다른 셀을 선택하면 새로운 메모 셀이 사라지는 동작이 구현하지 못했습니다.
> 해당 문제가 디테일뷰의 `func textViewDidEndEditing(_ textView: UITextView)` delegate 를 기점으로 테이블 데이터를
> 업데이트 하는데 FullScreen으로 한 화면씩 보이는 경우에서는 뒤로가기 탭으로 테이블뷰로 나오면
> 원하는 데로 되는데 위 사진처럼 할때만 생기는 문제였습니다..
> 
> 이렇다보니 제가 구현한 형태로 일일이 업데이트되는 시점을 찾고 테이블뷰가 refresh 되게 만드는 방향이 맞는건지..
> 스플릿 뷰에서 제가 놓친 기능이 있거나 다른 형태로 나누어진 View간에 통신 방법은 없는지 궁금합니다..
> 
> 혹시 NotificationCenter를 각 뷰간에 통신의 방법으로 사용하는 점도 어떤지 궁금합니다..
> 저희가 진행했던 쥬스메이커 만들기 공부할때 NotificationCenter 사용은 해당 프로젝트 상황(모델의 데이터 수정하는 요청)에 어울리지 않다고 피드백을 받았는데,
> 그러면 주로 어떤 상황에서 Notification Center가 사용되는 지 알고 싶습니다.
> 
> 사실 코드의 모든 부분을 질문을 하고 싶을 정도로 혼자 하다보니 다양한 선택과 동작 구현에만 초점이 잡힌 것 같습니다..
> 사소한 부분이라도 이상한 구현이나 잘못된 부분이 있다면 말씀해주세요 ㅎㅎ;;
> 
> 잘부탁드립니다!

👉 `우선 NotificationCenter의 경우는 어떤 동작을 여러곳의 화면이나, 클래스에서 받아 처리해야 하는 부분인 경우에 사용합니다. 물론 이러한 부분도 개인차는 다 있어요. 저의 경우는 최대한 그런 경우를 만들지 않으려고 하지만, 구조상 어쩔 수 없이 그렇게 되었다면 말씀드린대로 사용하는 편입니다 :) 다음은 메모셀을 날리는 부분에 대한건데요! 현재 코드에서는 메모를 선택하면 rootDelegate(네이밍좀..) 라는 메인 모델에 바로 인서트 해주는 형태를 취하고 있습니다. 그리고 스플릿 뷰의 디테일 화면을 넘겨주죠! 그 후에 secondaryViewController에서 작성된 데이터가 없는경우 진짜 Delegate Protocol 을 사용하거나, 클로저를 사용하여 데이터를 넘겨줄 수 있습니다. 개인적인 추천을 하고싶진 않지만 NotificationCenter 를 사용해도 데이터 전달은 가능하구요!`

말꼬 Step2 까지 하시느라 고생 많으셨어요. 하지만 데이터 전달에 대해서 어려움이 조금 있으신 것 같아요! Delegate Protocol만 잘 활용해 보시면 좀 더 좋은 코드가 나올 수 있을 것 같네요!
그리고 아이패드에서는 아이폰처럼 ActionSheet를 사용하는데에는 문제가 있습니다 ㅎㅎ 요 부분은 아이패드에서 실행해 보시면 크래시가 발생하니까, 해결하시는데 오랜 시간이 걸리진 않을 것 같아요!

많은 양의 코드 수정을 요하진 않고, 데이터를 주고받고 처리하는 부분만 Delegate 패턴을 이용하여 수정하는걸 한번 보고싶네요!

이해가 안되거나, 의도와 다른 답변이 달린경우 디코로 알려주세요!

Marcorable committed:
docs: 주석 삭제 및 SwiftLint 수정, CocoaPods 수정

Marcorable committed:
refactor: 문자열 enum 처리

Marcorable committed:
refactor: 앱 전반에 사용될 메모 배열 MemoCoreDataManager로 이동 및 CRUD 수정

CRUD 마치고 UI 변경을 위한 탈출 클로저 생성

Marcorable committed:
refactor: 배열 초기화 과정 MemoCoreDataManager로 이전

Marcorable committed:
refactor: Cell 그려주는 Configure 메서드 방식 변경

Marcorable committed:
refactor: Children ViewController간의 소통 방법 변경

UI변경에 대한 메시지만 전달하고 CoreData에 직접 접근해서 Model 변경

Marcorable committed:
refactor: SplitViewController 인스턴스 삭제

Marcorable committed:
refactor: 버튼에 lazy 선언 및 target self

Marcorable committed:
refactor: 아이패드에서 action sheet 팝업 위치 지정해주기

Marcorable commented:
웨더 리뷰 피드백 감사합니다!!

깊게 고민해보기보다 직접적으로 질문을 많이 드린것 같다는 후회를 했습니다..;;
피드백을 받고 보니 SplitViewController의 참조인스턴스를 delegate로 표기하기도 하고 이상한점이 많군요..
PR을 보내고 UML을 짜보면서 거미줄처럼 엉켜있는 모습을 보며 많이 느끼고 직접적으로 피드백을 통해 
잘못된 점을 많이 느낀것 같습니다..

이런 경험도 경험이 될진 모르겠지만 또 대 수술을 한번 거쳐보면서 피드백주신 해결법들을 이용해서 수정도 해보았습니다. 
깊게 공부를 하지는 못했지만 어떤 부분인지 어떤 키워드로 공부해야 할지 알게되어서 Step2 리 팩토링을 진행해보았습니다. 
어제 하루동안 Dropbox 인증까지 진행해보고 Step4에서 다루는 검색이나 AttributeText 부분도 욕심이 나서 조금씩 조사를 해보았었는데..

오늘 피드백을 받고 늦은 시각까지.. 피드백 받은 부분과 제가 느꼈던 부분, 해결책을 받은 부분들 모두 구현해보는 시간을 가져보았습니다..

또 많이 바뀌긴했는데 VC간에는 UI 관련 요청만 서로 전달하고 CoreData 싱글턴 객체에 직접 저장하는 방법으로 변경해보았습니다..;;
조언해주신 "Delegate Protocol 이나 클로저로 데이터 전달하기", "SplitView동작 사용, 네이밍 등 수정" 등등

이번 Step 2에서 아쉬운 부분들이 잘 반영 되었을 지 수정된 부분에 관해 피드백을 부탁드려도 괜찮을까요? ㅠㅠ;;

* Step 2 PR 코드 UML
<img width="1442" alt="Screen Shot 2021-09-15 at 11 38 45 PM" src="https://user-images.githubusercontent.com/24707229/133647057-6e8a1082-0a4d-42cf-9004-ba6515965def.png">

* Step 2 PR 피드백 받고 수정한 코드 UML
<img width="1701" alt="image" src="https://user-images.githubusercontent.com/24707229/133711817-6f9a78ab-a87b-4e25-b355-6654024b0baf.png">

Marcorable committed:
refactor: MemoCoreDataManager 를 싱글턴 참조가 아니라 Master 뷰에서 인스턴스 참조를 시켜봄..

Marcorable committed:
refactor: 텍스트의 변화를 일일이 체크하지 않음

Marcorable committed:
refactor: Sample 데이터 추가하는 동작 주석처리

Marcorable committed:
docs: Step 2 회고 초안

SungPyo commented:
> 웨더 리뷰 피드백 감사합니다!!
> 
> 깊게 고민해보기보다 직접적으로 질문을 많이 드린것 같다는 후회를 했습니다..;;
> 피드백을 받고 보니 SplitViewController의 참조인스턴스를 delegate로 표기하기도 하고 이상한점이 많군요..
> PR을 보내고 UML을 짜보면서 거미줄처럼 엉켜있는 모습을 보며 많이 느끼고 직접적으로 피드백을 통해
> 잘못된 점을 많이 느낀것 같습니다..
> 
> 이런 경험도 경험이 될진 모르겠지만 또 대 수술을 한번 거쳐보면서 피드백주신 해결법들을 이용해서 수정도 해보았습니다.
> 깊게 공부를 하지는 못했지만 어떤 부분인지 어떤 키워드로 공부해야 할지 알게되어서 Step2 리 팩토링을 진행해보았습니다.
> 어제 하루동안 Dropbox 인증까지 진행해보고 Step4에서 다루는 검색이나 AttributeText 부분도 욕심이 나서 조금씩 조사를 해보았었는데..
> 
> 오늘 피드백을 받고 늦은 시각까지.. 피드백 받은 부분과 제가 느꼈던 부분, 해결책을 받은 부분들 모두 구현해보는 시간을 가져보았습니다..
> 
> 또 많이 바뀌긴했는데 VC간에는 UI 관련 요청만 서로 전달하고 CoreData 싱글턴 객체에 직접 저장하는 방법으로 변경해보았습니다..;;
> 조언해주신 "Delegate Protocol 이나 클로저로 데이터 전달하기", "SplitView동작 사용, 네이밍 등 수정" 등등
> 
> 이번 Step 2에서 아쉬운 부분들이 잘 반영 되었을 지 수정된 부분에 관해 피드백을 부탁드려도 괜찮을까요? ㅠㅠ;;
> 
> * Step 2 PR 코드 UML
> 
> <img alt="Screen Shot 2021-09-15 at 11 38 45 PM" width="1442" src="https://user-images.githubusercontent.com/24707229/133647057-6e8a1082-0a4d-42cf-9004-ba6515965def.png">
> 
> * Step 2 PR 피드백 받고 수정한 코드 UML
> 
> <img alt="image" width="1701" src="https://user-images.githubusercontent.com/24707229/133711817-6f9a78ab-a87b-4e25-b355-6654024b0baf.png">

이번 프로젝트의 마지막 PR이 되겠네요. 

이번 프로젝트에서는 `테스트` 가 가장 부족하지 않았나 싶습니다.
`Terminating app due to uncaught exception 'NSGenericException', reason: 'UIPopoverPresentationController (<UIPopoverPresentationController: 0x12da393d0>) should have a non-nil sourceView or barButtonItem set before the presentation occurs.'` 와 같은 Crash 같은건 아이패드 시뮬레이터의 짧은 테스트만으로도 찾을 수 있으니까요!

Delegate로 수정 처리한 부분, 크래시가 나와서 제대로된 확인은 힘들었지만 Share 기능 추가. 메모 삭제나, 버튼 초기화시 나타난 문제에 대한 수정 등 수정된 부분들은 잘 처리하신 것 같아요. 
다만, `전체적으로 코드가 복잡해 보이는데 주석이 조금은 있었으면 좋겠습니다.` 리뷰어 개개인의 차이는 있겠지만 Share에 대한 에러나, Alert, Coredata 에 대한 에러도 전체 Error case를 정의해서 관리하시면 조금 더 깔끔한 코드가 나올 것 같고, viewdidLoad나 init시에 UI 설정이 따로 분리되었으면 좋겠어요. `Alert`을 예로들면 비슷한 동작들이 많이 퍼져있는데, 이런 부분 또한 한곳으로 모아서 처리하면 좋을 것 같네요! 그 외에도 `coreManager.insertData(MemoData("", "", Date().timeIntervalSince1970, nil))` 처럼 변수 한줄 추가가 아쉽거나, 마음에 안들어서 인자에 바로 넣는 경우 저러한 코드가 많아지면 가독성에 문제가 생기니 차라리 한줄 더 늘리고 가독성을 좋게 만드는걸 추천드려요. if나 guard도 마찬가지 입니다.

많은 기능을 개발하고 시도해 보는것은 매우 좋은 일이고 지금도 잘 하고 계시니까 그 부분을 조금 더 섬세하게 학습하시면 더욱 더 좋을 것 같아요!

Marco는 피드백도 좋고 학습능력도 매우 좋다고 생각해서 오픈마켓 같은 프로젝트(제가 많이 익숙한)를 오랜기간 같이하면 정말 많은것을 알려드릴 수 있을 것 같네요!
PR은 여기서 끝나지만 수정된 점이나 궁금한점 등 생긴다면 기간 이후에도 언제든지 물어보셔도 됩니다.

고생하셨습니다.

@Marco @Keeplo

Marcorable committed:
docs: Step 2 고민한 점 작성
